libraries/AP_Compass/AP_Compass_IST8310.cpp

```diff
@@ -0,0 +1,176 @@
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_IST8310::probe(Compass &compass,
+                                             AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    AP_Compass_IST8310 *sensor = new AP_Compass_IST8310(compass, std::move(dev));
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_IST8310::AP_Compass_IST8310(Compass &compass,
+                                     AP_HAL::OwnPtr<AP_HAL::Device> dev)
+    : AP_Compass_Backend(compass)
+    , _dev(std::move(dev))
+{
+}
+
+bool AP_Compass_IST8310::init()
+{
+    uint8_t val = 0;
+    bool ret;
+
+    hal.scheduler->suspend_timer_procs();
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf("IST8310: Unable to get bus semaphore\n");
+        goto fail_sem;
+    }
+
+    ret = _dev->read_registers(WAI_REG, &val, 1);
+    if (!ret) {
+        goto bus_error;
+    }
+    if (val != DEVICE_ID) {
+        hal.console->printf("IST8310: Wrong id\n");
+        goto fail;
+    }
+
+    ret = _dev->write_register(CNTL1_REG, SINGLE_MEASUREMENT_MODE);
+    if (!ret) {
+        goto bus_error;
+    }
+
+    _dev->get_semaphore()->give();
+    hal.scheduler->resume_timer_procs();
+
+    /* register the compass instance in the frontend */
+    _dev->set_device_type(DEVTYPE_IST8310);
+    set_dev_id(_compass_instance, _dev->get_bus_id());
+
+    _dev->register_periodic_callback(10 * USEC_PER_MSEC, FUNCTOR_BIND_MEMBER(&AP_Compass_IST8310::_update, bool));
+
+    return true;
+
+bus_error:
+    hal.console->printf("IST8310: Bus communication error\n");
+fail:
+    _dev->get_semaphore()->give();
+fail_sem:
+    hal.scheduler->resume_timer_procs();
+    return false;
+}
+
+bool AP_Compass_IST8310::_update()
+{
+    const auto now = AP_HAL::micros();
+
+    struct PACKED
+    {
+        uint8_t status;
+        le16_t rx;
+        le16_t ry;
+        le16_t rz;
+    } buffer;
+
+    bool ret = _dev->read_registers(STAT1_REG, (uint8_t *) &buffer, sizeof(buffer));
+
+    if (!ret) {
+        return true; /* We're going to be back on the next iteration either way */
+    }
+
+
+    auto status = buffer.status;
+
+    /* Make it compatible with HMC5883L */
+    auto x = - static_cast<int16_t>(le16toh(buffer.rx));
+    auto y = - static_cast<int16_t>(le16toh(buffer.ry));
+    auto z = - static_cast<int16_t>(le16toh(buffer.rz));
+
+    Vector3f raw_field = Vector3f{static_cast<float>(x),
+                                  static_cast<float>(y),
+                                  static_cast<float>(z)
+                                  };
+
+    /* convert uT to milligauss */
+    raw_field *= 3.0f;
+
+    /* rotate raw_field from sensor frame to body frame */
+    rotate_field(raw_field, _compass_instance);
+
+    /* publish raw_field (uncorrected point sample) for calibration use */
+    publish_raw_field(raw_field, now, _compass_instance);
```

Comment:
other sensors use the `publish time` instead of time in which we are running the callback. Main difference is that the latter doesn't include the I2C transfer time.